File: frontend/lib/features/comment/data/models/comment_dto.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:frontend/features/post/domain/entities/author.dart';
import '../../domain/entities/comment.dart';

part 'comment_dto.freezed.dart';
part 'comment_dto.g.dart';

/// Comment DTO (Data Transfer Object)
///
/// API 응답에서 받은 댓글 데이터를 Domain Entity로 변환하기 위한 중간 객체입니다.
@freezed
class CommentDto with _$CommentDto {
  const CommentDto._();

  const factory CommentDto({
    /// 댓글 고유 ID
    required int id,

    /// 게시글 ID
    required int postId,

    /// 댓글 내용
    required String content,

    /// 작성자 정보 (중첩된 객체)
    required Author author,

    /// 작성 시각
    required DateTime createdAt,

    /// 수정 시각 (선택적)
    DateTime? updatedAt,

    /// 댓글 깊이 (0=최상위, 1=대댓글)
    @Default(0) int depth,

    /// 부모 댓글 ID (선택적)
    int? parentCommentId,
  }) = _CommentDto;

  /// JSON에서 CommentDto 객체 생성
  factory CommentDto.fromJson(Map<String, dynamic> json) =>
      _$CommentDtoFromJson(json);

  /// DTO를 Domain Entity로 변환
  Comment toEntity() {
    return Comment(
      id: id,
      postId: postId,
      content: content,
      author: author,
      createdAt: createdAt,
      updatedAt: updatedAt,
      depth: depth,
      parentCommentId: parentCommentId,
    );
  }
}

File: backend_new/src/main/kotlin/com/univgroup/domain/content/dto/CommentDto.kt
package com.univgroup.domain.content.dto

import com.univgroup.domain.content.entity.Comment
import java.time.LocalDateTime

/**
 * 댓글 응답 DTO
 */
data class CommentDto(
    val id: Long,
    val postId: Long,
    val authorId: Long,
    val authorName: String,
    val authorProfileImageUrl: String?,
    val parentCommentId: Long?,
    val content: String,
    val likeCount: Long,
    val isDeleted: Boolean,
    val createdAt: LocalDateTime,
    val updatedAt: LocalDateTime,
    val replies: List<CommentDto>?,
) {
    companion object {
        fun from(
            comment: Comment,
            includeReplies: Boolean = false,
            replies: List<Comment> = emptyList(),
        ): CommentDto {
            return CommentDto(
                id = comment.id!!,
                postId = comment.post.id!!,
                authorId = comment.author.id!!,
                authorName = comment.author.name,
                authorProfileImageUrl = comment.author.profileImageUrl,
                parentCommentId = comment.parentComment?.id,
                content = comment.content,
                likeCount = comment.likeCount,
                isDeleted = comment.isDeleted,
                createdAt = comment.createdAt,
                updatedAt = comment.updatedAt,
                replies = if (includeReplies) replies.map { from(it) } else null,
            )
        }
    }
}

/**
 * 댓글 생성 요청 DTO
 */
data class CreateCommentRequest(
    val content: String,
    val parentCommentId: Long? = null,
)

/**
 * 댓글 수정 요청 DTO
 */
data class UpdateCommentRequest(
    val content: String,
)
